File: core/src/com/bombbird/terminalcontrol/TerminalControl.java
package com.bombbird.terminalcontrol;

import com.badlogic.gdx.Application;
import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.bombbird.terminalcontrol.entities.aircrafts.Emergency;
import com.bombbird.terminalcontrol.screens.RadarScreen;
import com.bombbird.terminalcontrol.sounds.TextToSpeech;
import com.bombbird.terminalcontrol.utilities.DiscordManager;
import com.bombbird.terminalcontrol.utilities.Fonts;
import com.bombbird.terminalcontrol.screens.MainMenuScreen;
import com.bombbird.terminalcontrol.utilities.RenameManager;
import com.bombbird.terminalcontrol.utilities.ToastManager;
import com.bombbird.terminalcontrol.utilities.saving.FileLoader;
import com.bombbird.terminalcontrol.utilities.saving.GameSaver;
import org.json.JSONObject;

public class TerminalControl extends Game {
    //Get screen size
    public static int WIDTH;
    public static int HEIGHT;

    //Is html? (freetype not supported in html)
    public static boolean ishtml;

    //Use discord? (initialisation successful?)
    public static boolean useDiscord;

    //Version info
    public static boolean full;
    public static String versionName;
    public static int versionCode;

    //Active gameScreen instance
    public static RadarScreen radarScreen = null;

    //Create texture stuff
    private static TextureAtlas buttonAtlas;
    public static Skin skin;

    //The one and only spritebatch
    public SpriteBatch batch;

    //Text-to-speech (for Android only)
    public static TextToSpeech tts;

    //Toast (for Android only)
    public static ToastManager toastManager;

    //Discord RPC (for Desktop only)
    public static boolean loadedDiscord;
    public static DiscordManager discordManager;

    //Default settings
    public static int trajectorySel;
    public static RadarScreen.Weather weatherSel;
    public static int soundSel;
    public static Emergency.Chance emerChance;
    public static boolean sendAnonCrash;
    public static boolean increaseZoom;
    public static int saveInterval;
    public static float radarSweep;
    public static int advTraj;
    public static int areaWarning;
    public static int collisionWarning;
    public static int defaultTabNo;
    public static int revision;
    public static final int LATEST_REVISION = 1;

    public TerminalControl(TextToSpeech tts, ToastManager toastManager, DiscordManager discordManager) {
        TerminalControl.tts = tts;
        TerminalControl.toastManager = toastManager;
        TerminalControl.discordManager = discordManager;

        loadedDiscord = false;
        useDiscord = false;
    }

    public static void loadSettings() {
        JSONObject settings = FileLoader.loadSettings();
        if (settings == null) {
            //Default settings if save unavailable
            trajectorySel = 90;
            weatherSel = RadarScreen.Weather.LIVE;
            if (Gdx.app.getType() == Application.ApplicationType.Desktop) {
                soundSel = 1;
            } else if (Gdx.app.getType() == Application.ApplicationType.Android) {
                soundSel = 2;
            }
            sendAnonCrash = true;
            increaseZoom = false;
            saveInterval = 60;
            radarSweep = 2;
            advTraj = -1;
            areaWarning = -1;
            collisionWarning = -1;
            defaultTabNo = 0;
            emerChance = Emergency.Chance.MEDIUM;
            revision = 0;
        } else {
            trajectorySel = settings.getInt("trajectory");
            String weather = settings.optString("weather");
            if ("true".equals(weather)) {
                //Old format
                weatherSel = RadarScreen.Weather.LIVE;
            } else if ("false".equals(weather)) {
                //Old format
                weatherSel = RadarScreen.Weather.RANDOM;
            } else {
                //New format
                weatherSel = RadarScreen.Weather.valueOf(settings.getString("weather"));
            }
            soundSel = settings.getInt("sound");
            sendAnonCrash = settings.optBoolean("sendCrash", true);
            if (settings.isNull("emerChance")) {
                emerChance = Emergency.Chance.MEDIUM;
            } else {
                emerChance = Emergency.Chance.valueOf(settings.getString("emerChance"));
            }
            increaseZoom = settings.optBoolean("increaseZoom", false);
            saveInterval = settings.optInt("saveInterval", 60);
            radarSweep = (float) settings.optDouble("radarSweep", 2);
            advTraj = settings.optInt("advTraj", -1);
            areaWarning = settings.optInt("areaWarning", -1);
            defaultTabNo = settings.optInt("defaultTabNo", 0);
            collisionWarning = settings.optInt("collisionWarning", -1);
            revision = settings.optInt("revision", 0);
        }
        GameSaver.saveSettings();
    }

    public static void loadVersionInfo() {
        String[] info = Gdx.files.internal("game/type.type").readString().split(" ");
        full = !"lite".equals(info[0]);
        versionName = info[1];
        versionCode = Integer.parseInt(info[2]);
        FileLoader.mainDir = TerminalControl.full ? "AppData/Roaming/TerminalControlFull" : "AppData/Roaming/TerminalControl";
    }

    public static boolean revisionNeedsUpdate() {
        return revision < LATEST_REVISION;
    }

    public static void updateRevision() {
        revision = LATEST_REVISION;
        GameSaver.saveSettings();
    }

    @Override
    public void create() {
        WIDTH = Gdx.graphics.getWidth();
        HEIGHT = Gdx.graphics.getHeight();
        batch = new SpriteBatch();

        buttonAtlas = new TextureAtlas(Gdx.files.internal("game/ui/mainmenubuttons.atlas"));
        skin = new Skin();
        skin.addRegions(TerminalControl.buttonAtlas);

        RenameManager.loadMaps();

        Gdx.input.setCatchBackKey(true);

        this.setScreen(new MainMenuScreen(this, null));
    }

    @Override
    public void render() {
        super.render();
    }

    @Override
    public void dispose() {
        batch.dispose();
        Fonts.dispose();
        buttonAtlas.dispose();
        skin.dispose();
    }

}

File: core/src/com/bombbird/terminalcontrol/TerminalControl.kt
package com.bombbird.terminalcontrol

import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Window
import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.entities.aircrafts.Emergency
import com.bombbird.terminalcontrol.screens.MainMenuScreen
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen
import com.bombbird.terminalcontrol.sounds.TextToSpeechInterface
import com.bombbird.terminalcontrol.sounds.TextToSpeechManager
import com.bombbird.terminalcontrol.utilities.*
import com.bombbird.terminalcontrol.utilities.RenameManager.loadMaps
import com.bombbird.terminalcontrol.utilities.files.ExternalFileHandler
import com.bombbird.terminalcontrol.utilities.files.FileLoader
import com.bombbird.terminalcontrol.utilities.files.GameSaver

class TerminalControl(tts: TextToSpeechInterface, toastManager: ToastManager, discordManager: DiscordManager, externalFileHandler: ExternalFileHandler, browserInterface: BrowserInterface, playServicesInterface: PlayServicesInterface) : Game() {
    companion object {
        //Get screen size
        var WIDTH = 0
        var HEIGHT = 0

        //Use discord? (initialisation successful?)
        var useDiscord = false

        //Version info
        const val full = true
        lateinit var versionName: String
        var versionCode = 0

        //Active gameScreen instance
        var radarScreen: RadarScreen? = null

        //Create texture stuff
        private lateinit var buttonAtlas: TextureAtlas
        lateinit var skin: Skin

        //Text-to-speech
        lateinit var ttsManager: TextToSpeechManager
        lateinit var tts: TextToSpeechInterface

        //Toast (for Android only)
        lateinit var toastManager: ToastManager

        //Discord RPC (for Desktop only)
        var loadedDiscord = false
        lateinit var discordManager: DiscordManager

        //External file loader
        lateinit var externalFileHandler: ExternalFileHandler

        //Browser interface
        lateinit var browserInterface: BrowserInterface

        //Play Games interface
        lateinit var playServicesInterface: PlayServicesInterface

        //Datatag config list
        var datatagConfigs = Array<String>()

        //Default settings
        var trajectorySel = 0
        var pastTrajTime = 0
        lateinit var weatherSel: RadarScreen.Weather
        var stormNumber = 0
        var soundSel = 0
        lateinit var emerChance: Emergency.Chance
        var sendAnonCrash = false
        var increaseZoom = false
        var saveInterval = 0
        var radarSweep = 0f
        var advTraj = 0
        var areaWarning = 0
        var collisionWarning = 0
        var showMva = false
        var showIlsDash = false
        var datatagConfig = "Default"
        var showUncontrolled = false
        var alwaysShowBordersBackground = false
        var rangeCircleDist = 0
        var lineSpacingValue = 0
        var colourStyle = 0
        var realisticMetar = false
        var distToGoVisible = 0
        var defaultTabNo = 0
        var revision = 0
        const val LATEST_REVISION = 1

        fun loadSettings() {
            val settings = FileLoader.loadSettings()
            if (settings == null) {
                //Default settings if save unavailable
                trajectorySel = 90
                pastTrajTime = -1
                weatherSel = RadarScreen.Weather.LIVE
                stormNumber = 0
                soundSel = 2
                sendAnonCrash = true
                increaseZoom = false
                saveInterval = 60
                radarSweep = 2f
                advTraj = -1
                areaWarning = -1
                collisionWarning = -1
                showMva = true
                showIlsDash = false
                datatagConfig = "Default"
                showUncontrolled = false
                alwaysShowBordersBackground = true
                rangeCircleDist = 0
                lineSpacingValue = 1
                colourStyle = 0
                realisticMetar = false
                distToGoVisible = 0
                defaultTabNo = 0
                emerChance = Emergency.Chance.MEDIUM
                revision = 0
            } else {
                trajectorySel = settings.optInt("trajectory", 90)
                pastTrajTime = settings.optInt("pastTrajTime", -1)
                val weather = settings.optString("weather")
                weatherSel = when (weather) {
                    "true" -> RadarScreen.Weather.LIVE //Old format
                    "false" -> RadarScreen.Weather.RANDOM //Old format
                    else -> RadarScreen.Weather.valueOf(settings.getString("weather")) //New format
                }
                soundSel = settings.optInt("sound", 2)
                stormNumber = settings.optInt("stormNumber", 0)
                sendAnonCrash = settings.optBoolean("sendCrash", true)
                emerChance = if (settings.isNull("emerChance")) {
                    Emergency.Chance.MEDIUM
                } else {
                    Emergency.Chance.valueOf(settings.getString("emerChance"))
                }
                increaseZoom = settings.optBoolean("increaseZoom", false)
                saveInterval = settings.optInt("saveInterval", 60)
                radarSweep = settings.optDouble("radarSweep", 2.0).toFloat()
                advTraj = settings.optInt("advTraj", -1)
                areaWarning = settings.optInt("areaWarning", -1)
                defaultTabNo = settings.optInt("defaultTabNo", 0)
                collisionWarning = settings.optInt("collisionWarning", -1)
                showMva = settings.optBoolean("showMva", true)
                showIlsDash = settings.optBoolean("showIlsDash", false)
                datatagConfig = settings.optString("datatagConfig", if (settings.optBoolean("compactData", false)) "Compact" else "Default")
                showUncontrolled = settings.optBoolean("showUncontrolled", false)
                alwaysShowBordersBackground = settings.optBoolean("alwaysShowBordersBackground", true)
                rangeCircleDist = settings.optInt("rangeCircleDist", 0)
                lineSpacingValue = settings.optInt("lineSpacingValue", 1)
                colourStyle = settings.optInt("colourStyle", 0)
                realisticMetar = settings.optBoolean("realisticMetar", false)
                distToGoVisible = settings.optInt("distToGoVisible", 0)
                revision = settings.optInt("revision", 0)
            }
            GameSaver.saveSettings()
        }

        fun loadVersionInfo() {
            val info = Gdx.files.internal("game/type.type").readString().split(" ".toRegex()).toTypedArray()
            versionName = info[0]
            versionCode = info[1].toInt()
            FileLoader.mainDir = if (full) "AppData/Roaming/TerminalControlFull" else "AppData/Roaming/TerminalControl"
        }

        fun revisionNeedsUpdate(): Boolean {
            return revision < LATEST_REVISION
        }

        fun updateRevision() {
            revision = LATEST_REVISION
            GameSaver.saveSettings()
        }

        fun loadDatatagConfigs() {
            datatagConfigs.clear()
            datatagConfigs.add("Default", "Compact")
            if (full) datatagConfigs.addAll(Array(FileLoader.getAvailableDatatagConfigs()))
        }
    }

    //The one and only spritebatch
    lateinit var batch: SpriteBatch

    init {
        Companion.tts = tts
        Companion.toastManager = toastManager
        Companion.discordManager = discordManager
        Companion.externalFileHandler = externalFileHandler
        Companion.browserInterface = browserInterface
        Companion.playServicesInterface = playServicesInterface
        loadedDiscord = false
        useDiscord = false
        ttsManager = TextToSpeechManager()
    }

    private fun loadDialogSkin() {
        val ws = Window.WindowStyle()
        ws.titleFont = Fonts.defaultFont20
        ws.titleFontColor = Color.WHITE
        ws.background = skin.getDrawable("ListBackground")
        ws.stageBackground = skin.getDrawable("DarkBackground")
        skin.add("defaultDialog", ws)
    }

    override fun create() {
        WIDTH = Gdx.graphics.width
        HEIGHT = Gdx.graphics.height
        batch = SpriteBatch()
        buttonAtlas = TextureAtlas(Gdx.files.internal("game/ui/terminal-control.atlas"))
        skin = Skin()
        skin.addRegions(buttonAtlas)
        loadDialogSkin()
        loadMaps()
        Gdx.input.setCatchKey(Input.Keys.BACK, true)
        setScreen(MainMenuScreen(this, null))
    }

    override fun dispose() {
        batch.dispose()
        Fonts.dispose()
        buttonAtlas.dispose()
        skin.dispose()
    }
}